packages/peer-exchange/src/waku_peer_exchange.ts

```diff
@@ -72,45 +57,43 @@ export class WakuPeerExchange implements IPeerExchange {
 
     const stream = await this.newStream(peer);
 
-    await pipe(
+    const res = await pipe(
       [rpcQuery.encode()],
       lp.encode(),
       stream,
       lp.decode(),
       async (source) => await all(source)
     );
-  }
 
-  /**
-   * Handle a peer exchange query response
-   */
-  private handler(streamData: IncomingStreamData): void {
-    const { stream } = streamData;
-    pipe(stream, lp.decode(), async (source) => {
-      for await (const bytes of source) {
-        const decoded = PeerExchangeRPC.decode(bytes).response;
-
-        if (!decoded) {
-          throw new Error("Failed to decode response");
-        }
-
-        const enrs = await Promise.all(
-          decoded.peerInfos.map(
-            (peerInfo) => peerInfo.enr && ENR.decode(peerInfo.enr)
-          )
-        );
-
-        const peerInfos = enrs.map((enr) => {
-          return {
-            ENR: enr,
-          };
-        });
-
-        if (!this.callback) throw new Error("Callback not set");
-
-        await this.callback({ peerInfos });
+    try {
+      const bytes = new Uint8ArrayList();
+      res.forEach((chunk) => {
+        bytes.append(chunk);
+      });
+
+      const decoded = PeerExchangeRPC.decode(bytes).response;
+
+      if (!decoded) {
+        throw new Error("Failed to decode response");
       }
-    }).catch((err) => log("Failed to handle peer exchange request", err));
+
+      const enrs = await Promise.all(
+        decoded.peerInfos.map(
+          (peerInfo) => peerInfo.enr && ENR.decode(peerInfo.enr)
+        )
+      );
+
+      const peerInfos = enrs.map((enr) => {
+        return {
+          ENR: enr,
+        };
+      });
```

Comment:
could probably merge both maps to avoid traversing the array twice

Reply:
```
      const peerInfos = await Promise.all(
        decoded.peerInfos.map(
          (peerInfo) => {
              if (!peerInfo.enr) return;
              
	          const enr = await ENR.decode(peerInfo.enr);

             return { ENR: enr }
          }
        )
      );
```

Probably need to use a typeguard:

```
export function isDefined<T>(msg: T | undefined): msg is T {
  return !!msg;
}
```

Or use this lib: https://twitter.com/mattpocockuk/status/1627686864579600385?s=20